aws-transfer-server/src/main/java/software/amazon/transfer/server/BaseHandlerStd.java

```diff
@@ -0,0 +1,361 @@
+            return translateToFailure(
+                    op, ex, HandlerErrorCode.InvalidRequest, model, ctx, reqToken);
+        }
+
+        return translateToFailure(
+                op, ex, HandlerErrorCode.GeneralServiceException, model, ctx, reqToken);
+    }
+
+    private ProgressEvent<ResourceModel, CallbackContext> translateToFailure(
+            String op,
+            Exception ex,
+            HandlerErrorCode errorCode,
+            ResourceModel model,
+            CallbackContext ctx,
+            String reqToken) {
+        logger.log(
+                String.format(FAILURE_LOG_MESSAGE, reqToken, model.getPrimaryIdentifier(), op, ex));
+        return ProgressEvent.failed(model, ctx, errorCode, ex.getMessage());
+    }
+
+    private boolean isRetryableServiceException(Exception exception) {
+        return exception instanceof ConflictException
+                || exception instanceof InternalServiceErrorException
+                || exception instanceof ServiceUnavailableException;
+    }
+
+    private ProgressEvent<ResourceModel, CallbackContext> handleRetryableServiceException(
+            ResourceModel model, CallbackContext ctx, String op, String reqToken, Exception ex) {
+
+        int currentNumThrottlingRetries = ctx.getNumThrottlingRetries();
+        if (currentNumThrottlingRetries > 0) {
+            ctx.setNumThrottlingRetries(currentNumThrottlingRetries - 1);
+            return ProgressEvent.defaultInProgressHandler(ctx, getDelaySeconds(ex), model);
+        }
+
+        HandlerErrorCode code;
+        if (ex instanceof ConflictException) {
+            code = HandlerErrorCode.ResourceConflict;
+        } else if (ex instanceof InternalServiceErrorException) {
+            code = HandlerErrorCode.ServiceInternalError;
+        } else if (ex instanceof ServiceUnavailableException) {
+            code = HandlerErrorCode.ServiceLimitExceeded;
+        } else {
+            code = HandlerErrorCode.GeneralServiceException;
+        }
+
+        return translateToFailure(op, ex, code, model, ctx, reqToken);
+    }
+
+    private boolean isThrottlingException(Exception e) {
+        if (StringUtils.contains(getErrorCode(e), THROTTLING_EXCEPTION_ERR_CODE)) {
+            return true;
+        }
+        if (e instanceof ThrottlingException) {
+            return true;
+        }
+        if (e instanceof AwsServiceException) {
+            return ((AwsServiceException) e).isThrottlingException();
+        }
+        return false;
+    }
+
+    private ProgressEvent<ResourceModel, CallbackContext> handleThrottling(
+            ResourceModel model, CallbackContext ctx, String op, String reqToken, Exception ex) {
+
+        int currentNumThrottlingRetries = ctx.getNumThrottlingRetries();
+        if (currentNumThrottlingRetries > 0) {
+            ctx.setNumThrottlingRetries(currentNumThrottlingRetries - 1);
+            return ProgressEvent.defaultInProgressHandler(ctx, getDelaySeconds(ex), model);
+        }
```

Comment:
All it means is that we should backoff and retry if necessary after a delay. I do not believe we should try and adjust the client retry policy. I prefer using the standard retry policies when possible and handle this here.

I picked this pattern from other handlers and I found it worth keeping.

Reply:
Why do we prefer to implement our own back off and retry mechanisms rather than relying on the SDK's built-in one?

Reply:
One reason is that we must respond to the calling state machine within 60 seconds or less. This mechanism is designed to use throttling based on standard retry strategy and then let the state machine do the overall retry so that it can remain in the IN_PROGRESS state. While we are retrying the rules for this chain reaction state machine dictate that we keep responding in a timely manner.

Reply:
Got it - responding within 60 seconds is a good point I was not aware of.